src/Services/ElementTabProvider.php

```diff
@@ -0,0 +1,123 @@
+<?php
+namespace DNADesign\Elemental\Services;
+
+use DNADesign\Elemental\Models\BaseElement;
+use Psr\SimpleCache\CacheInterface;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Forms\TabSet;
+
+/**
+ * Provides tab names for
+ */
+class ElementTabProvider implements Flushable
+{
+    use Configurable;
+
+    /**
+     * Set to true to refresh the tab cache on flush
+     *
+     * @config
+     * @var bool
+     */
+    private static $regenerate_on_flush = false;
+
+    /**
+     * @var CacheInterface
+     */
+    protected $cache;
+
+    /**
+     * ElementTabProvider constructor.
+     * @param CacheInterface $cache
+     */
+    public function __construct(CacheInterface $cache = null)
+    {
+        if (!$cache) {
+            $cache = Injector::inst()->get(CacheInterface::class . '.ElementTabCache');
+        }
+        $this->cache = $cache;
+    }
+
+    /**
+     * Get the top level tab names for the given element class
+     *
+     * @param string $elementClass
+     * @return array Array of the tabs for the element
+     */
+    public function getTabsForElement($elementClass)
+    {
+        if (null !== ($tabs = $this->cache->get($this->getCacheKey($elementClass)))) {
+            return $tabs;
+        }
+
+        return $this->generateTabsForElement($elementClass);
+    }
+
+    /**
+     * This function is triggered early in the request if the "flush" query
+     * parameter has been set. Each class that implements Flushable implements
+     * this function which looks after it's own specific flushing functionality.
+     *
+     * @see FlushMiddleware
+     */
+    public static function flush()
+    {
+        /** @var static $self */
+        $self = singleton(static::class);
+
+        $self->cache->clear();
+
+        if (self::config()->regenerate_on_flush) {
+            $self->generateAllTabs();
+        }
+    }
+
+    /**
+     * Identify and regenerate all tab names for all elemental blocks (and cache them)
+     *
+     * @return void
+     */
+    protected function generateAllTabs()
+    {
+        foreach (ClassInfo::subclassesFor(BaseElement::class) as $class) {
+            $this->generateTabsForElement($class);
+        }
+    }
+
+    /**
+     * Generate top level tab names for the given element class (and cache them)
+     *
+     * @param string $elementClass
+     * @return array
+     */
+    protected function generateTabsForElement($elementClass)
+    {
+        // We set an ID because some tabs only show if ->isInDb() is true
```

Comment:
Be careful: https://github.com/silverstripe/silverstripe-framework/pull/7799

Reply:
Hmmmn. This is just to get around `link_tracking` and `file_tracking` - which should hopefully not show by default soon. Might be able to remove it. Additionally it looks like that PR includes a way to "fake it" so we might be able to update this.

Reply:
I've removed the tabs in #374 because they break the ability set an element Title